crates/stages/stages/src/stages/merkle_changesets.rs

```diff
@@ -0,0 +1,374 @@
+
+        for block_number in target_range {
+            debug!(
+                target: "sync::stages::merkle_changesets",
+                ?block_number,
+                "Computing trie updates for block",
+            );
+            // Revert the state so that this block has been just processed, meaning we take the full
+            // revert of the subsequent block.
+            input.state = state_revert(block_number + 1);
+
+            // Construct prefix sets from only this block's `HashedPostState`, because we only care
+            // about trie updates which occurred as a result of this block being processed.
+            let this_state_revert = per_block_state_revert(block_number);
+            input.prefix_sets = this_state_revert.construct_prefix_sets();
+
+            // Calculate the trie updates for this block, then apply those updates to the reverts.
+            // We calculate the overlay which will be passed into the next step using the trie
+            // reverts prior to them being updated.
+            let this_trie_updates =
+                Self::calculate_block_trie_updates(provider, block_number, input.clone())?;
+
+            let trie_overlay = input.nodes.clone().into_sorted();
+            input.nodes.extend_ref(&this_trie_updates);
+            let this_trie_updates = this_trie_updates.into_sorted();
+
+            // Write the changesets to the DB using the trie updates produced by the block, and the
+            // trie reverts as the overlay.
+            debug!(
+                target: "sync::stages::merkle_changesets",
+                ?block_number,
+                "Writing trie changesets for block",
+            );
+            provider.write_trie_changesets(
+                block_number,
+                &this_trie_updates,
+                Some(&trie_overlay),
+            )?;
+        }
+
+        Ok(())
+    }
+}
+
+impl Default for MerkleChangeSets {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl<Provider> Stage<Provider> for MerkleChangeSets
+where
+    Provider:
+        StageCheckpointReader + TrieWriter + DBProvider + HeaderProvider + ChainStateBlockReader,
+{
+    fn id(&self) -> StageId {
+        StageId::MerkleChangeSets
+    }
+
+    fn execute(&mut self, provider: &Provider, input: ExecInput) -> Result<ExecOutput, StageError> {
+        // Get merkle checkpoint and assert that the target is the same.
+        let merkle_checkpoint = provider
+            .get_stage_checkpoint(StageId::MerkleExecute)?
+            .map(|checkpoint| checkpoint.block_number)
+            .unwrap_or(0);
+
+        if input.target.is_none_or(|target| merkle_checkpoint != target) {
+            return Err(StageError::Fatal(eyre::eyre!("Cannot sync stage to block {:?} when MerkleExecute is at block {merkle_checkpoint:?}", input.target).into()))
+        }
+
+        // Determine the target range for changeset computation
```

Comment:
not sure if this comment might be needed? seems quite intuitive from func already!